src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsEventLogForwarder.cs

```diff
@@ -0,0 +1,158 @@
+// <copyright file="SelfDiagnosticsEventLogForwarder.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using Microsoft.Extensions.Logging;
+
+namespace OpenTelemetry.Internal
+{
+    /// <summary>
+    /// SelfDiagnosticsEventListener class enables the events from OpenTelemetry event sources
+    /// and write the events to an ILoggerFactory.
+    /// </summary>
+    internal class SelfDiagnosticsEventLogForwarder : EventListener
```

Comment:
Is this intended for local debugging only, or for production use?
I'll be a bit concerned if it is for production:
1. The pattern matching "OpenTelemetry-*" and putting things in the `loggers` dictionary could be an attack surface, causing unbounded memory usage.
2. If folks consumes the logs in OpenTelemetry processor/exporters, it might dead loop (e.g. self-diagnostics go to logs, logs go to processor, processor triggered self-diagnostics).

Reply:
I would think it would just be for troubleshooting setups. Don't really think you'd ever want to have it on in production. Maybe briefly to figure out a telementry issue.

Reply:
#1258 has captured some of the requirements for a production-ready self-diagnostics solution.

If this feature is intended for non-production troubleshooting, is there a way we can clearly indicate that so people won't misuse it (e.g. a separate package, similar like the ConsoleExporter)? (e.g. in C++ folks normally do debug/release bits, doesn't seem to be a common practice in C# although technically it is supported)

Reply:
I mean, it's just listening to the event sources and writing to the logging system. Both of those things happen in production. I guess you might leave it on with a high minimum level like you do normal logging. So you can see errors and such.

Reply:
Also, @reyang we wouldn't need any of this if the whole project would just make use of MS.Ext.Logging like I was trying to argue for in that issue. That hasn't happened and the system currently uses event sources. So this is just forwarding those events to the standard logging system that every modern .NET app uses.

Reply:
And you know that we cannot use `Microsoft.Extensions.Logging` due to https://github.com/open-telemetry/opentelemetry-dotnet#supported-net-versions.

Reply:
yes. I remember Eric asking for ILogger in the originals PRs for self-diag.

Could we explore adding multiple destinations for SelfDiagnostic feature, File being one of them, ILogger being another (conditionally support where ILogger is supported).

Reply:
My biggest concern here is that we could easily end up with dead loop.

Reply:
@cijothomas that would require moving away from event sources being the primary thing that OTel is writing to internally.

Reply:
If a user has enabled OpenTelemetryLoggerProvider (eg: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/logs/getting-started/Program.cs#L33), then this would be a never-ending-loop.

Reply:
@ejsmith Could you check this please?
(Sorry I haven't yet reviewed the recent change, but this part was an open issue even before, and not sure if this was addressed).

Reply:
@cijothomas I just tool a look at this and the `Examples.AspNetCore` project already uses the OTel logging provider and it works fine. It does not result in a never-ending-loop.